base/sparse/sparsematrix.jl

```diff
@@ -120,7 +120,103 @@ column. In conjunction with [`nonzeros`](@ref) and
 """
 nzrange(S::SparseMatrixCSC, col::Integer) = S.colptr[col]:(S.colptr[col+1]-1)
 
+
+module SparseArrayValidity
+    @enum(_SparseArrayValidity,
+        VALID,
+        COLPTR_LENGTH,
+        COLPTR_FIRST_VAL,
+        NZVAL_LENGTH,
+        COLPTR_SORTED,
+        ROWVAL_LENGTH,
+        ROWVAL_RANGE,
+        ROWVAL_SORTED_COLUMN,
+    )
+end
+
+
+"""
+    checkvalid(Bool, S::SparseMatrixCSC; full = true) -> Bool
+
+Return `true` if the sparse matrix `S` has a valid internal representation.
+If `full` is `false` only O(1) checks are done, otherwise full
+O(nnz) checks are performed.
+
+# Examples
+
+```jldoctest
+julia> Base.SparseArrays.checkvalid(Bool, sprand(5,5,0.5))
+true
+
+julia> Base.SparseArrays.checkvalid(Bool, SparseMatrixCSC(4, 4, Int[], Int[], Int[]))
+false
+```
+"""
+function checkvalid(::Type{Bool}, S::SparseMatrixCSC; full = true)
+    return _checkvalid(S; full = full) == SparseArrayValidity.VALID
+end
+
+"""
+    checkvalid(S::SparseMatrixCSC; full = true)
+
+Throw an [`ArgumentError`](@ref) if the sparse matrix `S` does not have a valid internal representation.
+If `full` is `false` only O(1) checks are done, otherwise full
+O(nnz) checks are performed.
+"""
+function checkvalid(S::SparseMatrixCSC; full = true)
+    validity = _checkvalid(S; full = full)
+    validity == SparseArrayValidity.VALID && return
+    if validity == SparseArrayValidity.COLPTR_LENGTH
+        err_str = "length of S.colptr must be at least size(S,2) + 1 = $(S.n + 1) but was $(length(S.colptr))"
+    elseif validity == SparseArrayValidity.COLPTR_FIRST_VAL
+        err_str = "first element of S.colptr must be 1 but was $(S.colptr[1])"
+    elseif validity == SparseArrayValidity.ROWVAL_LENGTH
+        err_str = "length of S.rowval must be at least S.colptr[S.n+1] - 1 but was $(length(S.rowval))"
+    elseif validity == SparseArrayValidity.NZVAL_LENGTH
+        err_str = "length of S.nzval must be at least S.colptr[S.n+1] - 1 but was $(length(S.nzval))"
+    elseif validity == SparseArrayValidity.COLPTR_SORTED
+        err_str = "S.colptr must be sorted"
+    elseif validity == SparseArrayValidity.ROWVAL_RANGE
+        err_str = "each element in S.rowval must be > 0 and ⩽ size(S,1)"
+    elseif validity == SparseArrayValidity.ROWVAL_SORTED_COLUMN
+        err_str = "S.rowval must be sorted for all column ranges"
+    else # Should not happen
+        err_str = "sparse array is invalid"
+    end
```

Comment:
Regarding quoting error messages; do we usually quote things in error message strings? They will not be rendered as markdown and I'm not sure if we use "`" as a convention for code in strings.